donkeycar/parts/kinematics.py

```diff
@@ -1,12 +1,555 @@
+    takes the steering angle in radians and output of the odometer 
+    and turns those into:
+    - forward distance and velocity,
+    - pose; angle aligned (x,y) position and orientation in radians
+    - pose velocity; change in angle aligned position and orientation per second
+    @param wheel_base: distance between the front and back wheels
+
+    NOTE: this version uses the point midway between the rear wheels
+          as the point of reference.
+    see https://thef1clan.com/2020/09/21/vehicle-dynamics-the-kinematic-bicycle-model/
+    """
+    def __init__(self, wheel_base:float, debug=False):
+        self.wheel_base:float = wheel_base
+        self.debug = debug
+        self.timestamp:float = 0
+        self.forward_distance:float = 0
+        self.forward_velocity:float = 0
+        self.pose = Pose2D()
+        self.pose_velocity = Pose2D()
+        self.running:bool = True
+
+    def run(self, forward_distance:float, steering_angle:float, timestamp:float=None) -> Tuple[float, float, float, float, float, float, float, float, float]:
+        """
+        params
+            forward_distance: distance the reference point has travelled
+            steering_angle: angle in radians of the front 'wheel' from forward.
+                            In this case left is positive, right is negative,
+                            and directly forward is zero.
+            timestamp: time of distance readings or None to use current time
+        returns
+            distance
+            velocity
+            x is horizontal position of point midway between wheels
+            y is vertical position of point midway between wheels
+            angle is orientation in radians around point midway between wheels
+            x' is the horizontal velocity
+            y' is the vertical velocity
+            angle' is the angular velocity
+            timestamp
+
+        """
+        if timestamp is None:
+            timestamp = time.time()
+
+        steering_angle = limit_angle(steering_angle)
+
+        if self.running:
+            if 0 == self.timestamp:
+                self.forward_distance = forward_distance
+                self.forward_velocity=0
+                self.pose = Pose2D()
+                self.pose_velocity = Pose2D()
+                self.timestamp = timestamp
+            elif timestamp > self.timestamp:
+                #
+                # changes from last run
+                #
+                delta_time = timestamp - self.timestamp
+                delta_distance = forward_distance - self.forward_distance
+                forward_velocity = delta_distance / delta_time
+
+                #
+                # new velocities
+                #
+                angle_velocity = bicycle_angular_velocity(self.wheel_base, forward_velocity, steering_angle)
+                delta_angle = angle_velocity * delta_time
+                estimated_angle = limit_angle(self.pose.angle + delta_angle / 2)
```

Comment:
The process assumes very small amount of time between updates such that it can assume the car has move along a line that inverval.  It need to know the orientation of the vehicle during the inverval.  At the start of the interval is was self.pose.angle, by the end of the interval it was self.pose.angle + delta_angle.  So we use the average of these two values for the angle of travel during the interval.

Reply:
Ok, thanks, makes sense.